_posts/2019-12-26-policy-gradients.md

```diff
@@ -0,0 +1,136 @@
+3. Learned baselines: $$\inline b(s_{t}) = V_{\phi}(s_{t}) $$ (these kinds have a separate critic, which is, in a sense, evaluating the agent’s actions. Discussed next)
+
+
+## Actor-Critic methods
+
+
+These are policy gradient methods where the baseline is also learned.
+Two networks: Critic $$\inline V_{\phi}(s)$$ and Actor $$\inline \pi_{\theta}(a|s)$$
+Both may be separate networks, or may share initial layers and branch out at a later layer.
+
+Critic: did this state turn out more rewarding than I expected?
+
+Actor: did this action send us into a better situation than we thought it would?
+
+If we use this value function as a baseline, our gradient term becomes: 
+
+$$\nabla_{\theta} U(\pi)\ \approx {1 \over m}\sum_{i=0}^{m-1}\sum_{t=0}^{H-1} (RTG(\tau, t) - V(s_{t})) \nabla_{\theta}\ log\ \pi (a_{t}|s_{t})$$
+
+There's something to notice here. For the gradient, along with the grad log probability term, we are using a _weight_ multiplier, to indicate whether we want to increase (+ve weight) or decrease (-ve weight) the probability of that action in that state.
+
+The notion of “better than average” is captured nicely by the _Advantage_ function:
+
+$$A(s, a) = Q(s, a) - V(s)$$
+
+
+
+$$\nabla_{\theta} U(\pi)\ \approx {1 \over m}\sum_{i=0}^{m-1}\sum_{t=0}^{H-1} (Q(s_{t}^{(i)}, a_{t}^{(i)}) - V(s_{t}^{(i)})) \nabla_{\theta}\ log\ \pi (a_{t}^{(i)}|s_{t}^{(i)})$$
+
+Using Q-values instead of raw returns helps to reduce variance, since the Q-value captures information from multiple branches following this $$\inline (s, a)$$ choice, whereas $$\inline R(\tau)$$ only gives information about one path.
+
+Now the problem is, we don't have an explicit Q-network, so we'll need to estimate the Q-values somehow. One simple way to estimate A is using the _TD error_:
+
+$$A(s_{t}, a_{t}) = r_{t} + V(s_{t+1}) - V(s_{t})$$
+
+The same TD error serves two purposes: as the advantage function for the actor, and as the value error for the critic.
+
+The A3C paper takes A(s, a) as
+
+$$A(s_t, a_t) = [r_t + \gamma ^1r_{t+1} + … + \gamma ^kr_{t+k} + V(s_{t+k+1})] - V(s_t)$$
+
+(for some fixed k, or until terminal state, whichever comes first. $$\inline \gamma$$ is the discount factor)
+
+### Sources
+* [Spinning up in Deep RL by OpenAI](https://spinningup.openai.com)
+* [Deep RL Bootcamp by Deepmind](https://sites.google.com/view/deep-rl-bootcamp/lectures)
+* [Daniel Takeshi's blog](https://danieltakeshi.github.io/). Specifically, [this post](https://danieltakeshi.github.io/2017/03/28/going-deeper-into-reinforcement-learning-fundamentals-of-policy-gradients/) and [this post](https://danieltakeshi.github.io/2018/06/28/a2c-a3c/)
+* [Lilian Weng's blog](https://lilianweng.github.io/). [This post](https://lilianweng.github.io/lil-log/2018/04/08/policy-gradient-algorithms.html) in particular.
+
+_(I plan on adding more algorithms as soon as I understand them a bit better. TRPO, PPO, SAC, DDPG, etc.)_
+
```

Comment:
Overall well written! It is informative, though not very beginner-friendly. Adding some more explanation would make it better. Otherwise, it looks good to me.